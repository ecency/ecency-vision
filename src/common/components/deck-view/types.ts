export interface DeckModel {
  data: any;
  dataParams: any[];
  header: {
    title: string;
    icon: JSX.Element | null;
  };
  listItemComponent: any;
  createdAt: Date;
}

export interface IdentifiableDeckModel extends DeckModel {
  id: string;
  content: string;
}
