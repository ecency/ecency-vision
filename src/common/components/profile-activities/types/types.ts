export interface ActivityTypes {
    type: string;
      parent_author?: string;
      parent_permlink?: string;
      author?: string;
      permlink?: string;
      payment?: number;
      payer?: string;
      following?: string;
      community?: string;
      approve?: boolean;
      proposal_ids?: string;
      account?: string;
      id?: string;
      what?: string[];
      voter?: string;
      witness?: string;
      json: string;
    timestamp: string;
  }