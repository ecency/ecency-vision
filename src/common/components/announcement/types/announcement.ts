export interface Announcement {
  id: number;
  title: string;
  description: string;
  button_text: string;
  button_link: string;
}

export interface LaterAnnouncement {
  id: number;
  dateTime: Date;
}
