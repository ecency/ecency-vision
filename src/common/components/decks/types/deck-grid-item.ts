interface BaseDeckGridItem {
  id: string;
  key: number;
  type:
    | "u"
    | "ac"
    | "co"
    | "w"
    | "n"
    | "tr"
    | "to"
    | "s"
    | "cu"
    | "th"
    | "msf"
    | "faq"
    | "wb"
    | "wn";
  settings: Record<string, any>;
}

export interface ReloadableDeckGridItem extends BaseDeckGridItem {
  settings: {
    updateIntervalMs: number;
  };
}

export interface UserDeckGridItem extends ReloadableDeckGridItem {
  type: "u";
  settings: ReloadableDeckGridItem["settings"] & {
    username: string;
    contentType: string;
  };
}

export interface CommunityDeckGridItem extends ReloadableDeckGridItem {
  type: "co";
  settings: UserDeckGridItem["settings"] & {
    tag: string;
  };
}

export interface SearchDeckGridItem extends ReloadableDeckGridItem {
  type: "s";
  settings: ReloadableDeckGridItem["settings"] & {
    query: string;
    author?: string;
    type?: string;
    category?: string;
    tags?: string;
    date?: string;
    sort?: string;
    hideLow?: boolean;
  };
}

export interface WavesDeckGridItem extends ReloadableDeckGridItem {
  type: "th";
  settings: ReloadableDeckGridItem["settings"] & {
    host: string;
  };
}

export type DeckGridItem =
  | BaseDeckGridItem
  | UserDeckGridItem
  | CommunityDeckGridItem
  | SearchDeckGridItem
  | WavesDeckGridItem;
