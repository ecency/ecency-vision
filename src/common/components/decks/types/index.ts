export * from "./deck-grid-item";
export * from "./deck-grid";
export * from "./old-decks";
