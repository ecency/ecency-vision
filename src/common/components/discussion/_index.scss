@import "src/style/vars_mixins";

.discussion {
  margin: 10px 0 30px 0;
  position: relative;
  min-height: 200px;

  @include themify(day) {
    border-top: 1px solid;
    @apply border-light-300;
  }

  @include themify(night) {
    border-top: 1px solid;
    @apply border-dark-200;
  }

  &.empty {
    border: none !important;
  }

  &.loading {
    border-color: transparent;
  }

  // fix progress bar to top
  .linear-progress {
    position: absolute;
    left: 0;
    top: 0;
  }

  .discussion-card {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding-top: 20px;

    .icon {
      padding: 14px;
      border-radius: 50%;
      margin-bottom: 20px;
      @apply text-blue-dark-sky;
      border: 2px solid;
      @apply border-blue-dark-sky;
      width: 54px;
      height: 54px;

      svg {
        height: 24px;
        margin-bottom: -2px;
      }
    }

    .label {
      font-size: 1.4rem;
      margin-bottom: 20px;
    }
  }

  .discussion-header {
    align-items: center;
    @apply text-gray-steel;
    display: flex;
    justify-content: space-between;
    margin-bottom: 15px;
    padding-top: 15px;
    font-size: 90%;

    @media (min-width: $sm-break) {
      font-size: initial;
    }

    .count {
      align-items: center;
      display: flex;

      svg {
        height: 16px;
        margin-right: 6px;
        opacity: 0.5;
      }
    }

    .order {
      display: flex;
      align-items: center;

      .order-label {
        white-space: nowrap;
        margin-right: 6px;
        display: none;
        @media (min-width: $sm-break) {
          display: block;
        }
      }
    }
  }

  .discussion-list {
    .discussion-item {

      &.selected-item > .item-inner {
        @apply bg-blue-dark-sky-040;
        border: 1px solid;
        @apply border-blue-dark-sky;
        padding-left: 4px;
        @include themify(night) {
          @apply bg-dark-200-010-light;
        }
      }

      .item-anchor {
        position: absolute;
        left: 10px;
        top: -90px;
        opacity: 0;
        width: 1px;
        height: 1px;
      }

      .item-inner {
        display: flex;
        padding: 12px 0;

        .item-figure {
          flex-grow: 0;
          flex-shrink: 0;
          margin-right: 10px;

          .user-avatar {
            width: 40px;
            height: 40px;

            @media (min-width: $md-break) {
              width: 40px;
              height: 40px;
            }
          }
        }

        .item-content {
          flex-grow: 1;
          font-size: 13px;
          width: calc(100% - 34px);

          @media (min-width: $md-break) {
            font-size: initial;
            width: calc(100% - 50px);
          }

          .item-header {
            display: flex;
            flex-direction: column;

            @media (min-width: $md-break) {
              align-items: center;
              flex-direction: row;
            }

            .author {
              align-items: center;
              display: inline-flex;
              padding: 2px;

              .author-name {
                font-weight: 700;
                word-break: keep-all;
              }

              @include themify(day) {
                @apply text-gray-charcoal;
              }

              @include themify(night) {
                @apply text-gray-pinkish;
              }

              .author-reputation {
                @apply text-gray-steel;
                font-weight: 500;

                &::before {
                  content: "(";
                }

                &::after {
                  content: ")";
                }
              }
              &:hover {
                @include themify(day) {
                  @apply bg-blue-duck-egg;
                }
                @include themify(night) {
                  @apply bg-blue-metallic;
                }
              }
            }

            .separator {
              display: none;
              @media (min-width: $md-break) {
                display: block;
              }
            }

            .date {
              opacity: 0.7;
            }
          }

          .reveal-item {
            margin-top: 12px;
          }

          .item-body {
            margin: 10px 0 14px 0;

            @include themify(day) {
              @apply text-tundora;
            }
          }

          .item-controls {
            align-items: center;
            display: flex;
            width: 100%;

            .reply-btn, .edit-btn, .delete-btn, .mute-btn {
              @apply text-gray-steel-010;
              cursor: pointer;

              &:hover {
                @apply text-gray-steel;
              }

              &.disabled {
                opacity: 0.4;
                cursor: not-allowed;
              }
            }

            .edit-btn, .delete-btn, .mute-btn {
              margin-left: 10px;
              @apply text-gray-steel-010;

              @media (min-width: $md-break) {
                margin-left: 16px;
              }

              svg {
                height: 14px;
              }
            }

            .delete-btn {
              &.in-progress {
                animation: anim-fadein-out 2s infinite;
                cursor: default;
              }
            }
          }

          .read-more {
            margin-top: 14px;
          }
        }
      }

      &.hidden-item {
        .item-inner {
          opacity: 0.6;
        }
      }

      .comment-box {
        padding-left: 20px;
        border-bottom: none;
      }

      .discussion-list {
        @include clearfix();
        margin-left: 18px;

        @media (min-width: $sm-break) {
          margin-left: 24px;
        }

        @media (min-width: $md-break) {
          margin-left: 50px;
        }

        @media (min-width: $md-break) {
          .discussion-item {
            // Add extra left padding and guide lines to sub items
            border-left: 1px dotted transparent;

            .item-inner {
              padding-left: 13px;
            }

            @include themify(day) {
              @apply border-light-400;
            }
            @include themify(night) {
              @apply border-dark-600-010-light;
            }
          }
        }
      }
    }
  }
  .profile-container-top {
    @media (min-width: $md-break) {
      width: 45%;
      top: -50px;
      left: -20px
    }
  }
  .profile-container-bottom {
    @media (min-width: $md-break) {
      width: 45%;
      bottom: 50px;
      left: -20px
    }
  }
}

.author-down-arrow {
  display: none;
  @media (max-width: $sm-break) {
    display: inline-flex;
    justify-content: center;
    width: 16px;
    height: 16px;
    align-items: center;
  }
}

.pointer {
  cursor: pointer;
}

.flex-1 {
  flex: 1
}

.dropdown-container{
  .dropdown-btn {
    color: #939a9f !important
  }
}
