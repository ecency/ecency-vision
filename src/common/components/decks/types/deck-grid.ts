import { DeckGridItem } from "./deck-grid-item";

export interface DeckGrid {
  key: string;
  title: string;
  icon?: string;
  columns: DeckGridItem[];
  storageType: "account" | "local";
}

export interface DeckGrids {
  decks: DeckGrid[];
}
