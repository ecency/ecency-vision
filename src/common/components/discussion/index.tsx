import React, {Component} from "react";

import {History, Location} from "history";

import moment from "moment";

import {Button, Form, FormControl} from "react-bootstrap";

import defaults from "../../constants/defaults.json";

import {renderPostBody, setProxyBase} from "@ecency/render-helper";

setProxyBase(defaults.imageServer);

import {Entry, EntryVote} from "../../store/entries/types";
import {Account, FullAccount} from "../../store/accounts/types";
import {Community, ROLES} from "../../store/communities/types";
import {DynamicProps} from "../../store/dynamic-props/types";
import {Global} from "../../store/global/types";
import {User} from "../../store/users/types";
import {ActiveUser} from "../../store/active-user/types";
import {Discussion as DiscussionType, SortOrder} from "../../store/discussion/types";
import {UI, ToggleType} from "../../store/ui/types";

import BaseComponent from "../base";
import ProfileLink from "../profile-link";
import EntryLink from "../entry-link";
import UserAvatar from "../user-avatar";
import EntryVoteBtn from "../entry-vote-btn/index";
import EntryPayout from "../entry-payout/index";
import EntryVotes from "../entry-votes";
import LinearProgress from "../linear-progress";
import Comment from "../comment"
import EntryDeleteBtn from "../entry-delete-btn";
import MuteBtn from "../mute-btn";
import LoginRequired from "../login-required";

import parseDate from "../../helper/parse-date";

import {_t} from "../../i18n";

import {comment, formatError} from "../../api/operations";

import * as ls from "../../util/local-storage";

import {createReplyPermlink, makeJsonMetaDataReply} from "../../helper/posting";
import tempEntry from "../../helper/temp-entry";

import {error} from "../feedback";

import _c from "../../util/fix-class-names"

import {commentSvg, pencilOutlineSvg, deleteForeverSvg} from "../../img/svg";

import {version} from "../../../../package.json";
import accountReputation from '../../helper/account-reputation';


interface ItemBodyProps {
    entry: Entry;
    global: Global;
}

export class ItemBody extends Component<ItemBodyProps> {
    shouldComponentUpdate(nextProps: Readonly<ItemBodyProps>): boolean {
        return this.props.entry.body !== nextProps.entry.body;
    }

    render() {
        const {entry, global} = this.props;

        const renderedBody = {__html: renderPostBody(entry.body, false, global.canUseWebp)};

        return <div className="item-body markdown-view mini-markdown" dangerouslySetInnerHTML={renderedBody}/>
    }
}


interface ItemProps {
    history: History;
    location: Location;
    global: Global;
    dynamicProps: DynamicProps;
    users: User[];
    activeUser: ActiveUser | null;
    discussion: DiscussionType;
    entry: Entry;
    community: Community | null;
    ui: UI;
    addAccount: (data: Account) => void;
    setActiveUser: (username: string | null) => void;
    updateActiveUser: (data?: Account) => void;
    deleteUser: (username: string) => void;
    updateReply: (reply: Entry) => void;
    addReply: (reply: Entry) => void;
    deleteReply: (reply: Entry) => void;
    toggleUIProp: (what: ToggleType) => void;
}

interface ItemState {
    reply: boolean;
    edit: boolean;
    inProgress: boolean;
    showIfHidden: boolean;
}

export class Item extends BaseComponent<ItemProps, ItemState> {
    state: ItemState = {
        reply: false,
        edit: false,
        inProgress: false,
        showIfHidden: false
    }

    afterVote = (votes: EntryVote[], estimated: number) => {
        const {entry, updateReply} = this.props;

        const {payout} = entry;
        const newPayout = payout + estimated;

        updateReply({
            ...entry,
            active_votes: votes,
            payout: newPayout,
            pending_payout_value: String(newPayout)
        });
    }

    toggleReply = () => {
        const {reply, edit} = this.state;
        if (edit) {
            return;
        }
        this.stateSet({reply: !reply});
    }

    toggleEdit = () => {
        const {edit, reply} = this.state;
        if (reply) {
            return;
        }
        this.stateSet({edit: !edit});
    }

    replyTextChanged = (text: string) => {
        const {entry} = this.props;
        ls.set(`reply_draft_${entry.author}_${entry.permlink}`, text);
    }

    submitReply = (text: string) => {
        const {entry} = this.props;
        const {activeUser, addReply, updateReply} = this.props;

        if (!activeUser || !activeUser.data.__loaded) {
            return;
        }

        const {author: parentAuthor, permlink: parentPermlink} = entry;
        const author = activeUser.username;
        const permlink = createReplyPermlink(entry.author);

        const jsonMeta = makeJsonMetaDataReply(
            entry.json_metadata.tags || ['ecency'],
            version
        );

        this.stateSet({inProgress: true});

        comment(
            author,
            parentAuthor,
            parentPermlink,
            permlink,
            '',
            text,
            jsonMeta,
            null,
            true
        ).then(() => {
            const nReply = tempEntry({
                author: activeUser.data as FullAccount,
                permlink,
                parentAuthor,
                parentPermlink,
                title: '',
                body: text,
                tags: []
            });

            // add new reply to store
            addReply(nReply);

            // remove reply draft
            ls.remove(`reply_draft_${entry.author}_${entry.permlink}`);

            // close comment box
            this.toggleReply();

            if (entry.children === 0) {
                // Update parent comment.
                const nParentReply: Entry = {
                    ...entry,
                    children: 1
                }

                updateReply(nParentReply);
            }
        }).catch((e) => {
            error(formatError(e));
        }).finally(() => {
            this.stateSet({inProgress: false});
        });
    }

    updateReply = (text: string) => {
        const {entry} = this.props;
        const {activeUser, updateReply} = this.props;

        const {permlink, parent_author: parentAuthor, parent_permlink: parentPermlink} = entry;
        const jsonMeta = makeJsonMetaDataReply(
            entry.json_metadata.tags || ['ecency'],
            version
        );

        this.stateSet({inProgress: true});

        comment(
            activeUser?.username!,
            parentAuthor!,
            parentPermlink!,
            permlink,
            '',
            text,
            jsonMeta,
            null,
        ).then(() => {
            const nReply: Entry = {
                ...entry,
                body: text
            }

            updateReply(nReply); // update store
            this.toggleEdit(); // close comment box
        }).catch((e) => {
            error(formatError(e));
        }).finally(() => {
            this.stateSet({inProgress: false});
        });
    }

    deleted = () => {
        const {entry, deleteReply} = this.props;
        deleteReply(entry);
    }

    render() {
        const {entry, activeUser, community, location} = this.props;
        const {reply, edit, inProgress, showIfHidden} = this.state;

        const created = moment(parseDate(entry.created));
        const reputation = accountReputation(entry.author_reputation);
        const readMore = entry.children > 0 && entry.depth > 5;
        const showSubList = !readMore && entry.children > 0;
        const canEdit = activeUser && activeUser.username === entry.author;

        const canMute = activeUser && community ? !!community.team.find(m => {
            return m[0] === activeUser.username &&
                [ROLES.OWNER.toString(), ROLES.ADMIN.toString(), ROLES.MOD.toString()].includes(m[1])
        }) : false;

        const isHidden = !!entry.stats?.gray && !showIfHidden;

        const anchorId = `anchor-@${entry.author}/${entry.permlink}`;

        const selected = location.hash && location.hash.replace("#", "") === `@${entry.author}/${entry.permlink}`;

        return (
            <div className={_c(`discussion-item depth-${entry.depth} ${isHidden ? "hidden-item" : ""} ${selected ? "selected-item" : ""}`)}>
                <div className="item-anchor" id={anchorId}/>
                <div className="item-inner">
                    <div className="item-figure">
                        {ProfileLink({...this.props, username: entry.author, children: <a>{UserAvatar({...this.props, username: entry.author, size: "medium"})}</a>})}
                    </div>
                    <div className="item-content">
                        <div className="item-header">
                            {ProfileLink({
                                ...this.props,
                                username: entry.author,
                                children: <div className="author notranslate">
                                    <span className="author-name">{entry.author}</span>
                                    <span className="author-reputation">{reputation}</span>
                                </div>
                            })}
                            <span className="separator"/>
                            {EntryLink({
                                ...this.props,
                                entry,
                                children: <span className="date" title={created.format("LLLL")}>
                                  {created.fromNow()}
                                </span>
                            })}
                        </div>
                        {(() => {
                            if (isHidden) {
                                return <div className="reveal-item">
                                    <Button variant="outline-danger" size="sm" onClick={() => {
                                        this.stateSet({showIfHidden: true});
                                    }}>{_t("discussion.reveal")}</Button>
                                </div>
                            }

                            return <>
                                <ItemBody global={this.props.global} entry={entry}/>
                                <div className="item-controls">
                                    {EntryVoteBtn({
                                        ...this.props,
                                        entry,
                                        afterVote: this.afterVote
                                    })}
                                    {EntryPayout({
                                        ...this.props,
                                        entry
                                    })}
                                    {EntryVotes({
                                        ...this.props,
                                        entry
                                    })}
                                    <a className={_c(`reply-btn ${edit ? 'disabled' : ''}`)} onClick={this.toggleReply}>
                                        {_t("g.reply")}
                                    </a>
                                    {(community && canMute) && MuteBtn({
                                        entry,
                                        community: community!,
                                        activeUser: activeUser!,
                                        onSuccess: (entry) => {
                                            const {updateReply} = this.props;
                                            updateReply(entry);
                                        }
                                    })}
                                    {canEdit && (
                                        <>
                                            <a title={_t('g.edit')} className={_c(`edit-btn ${reply ? 'disabled' : ''}`)} onClick={this.toggleEdit}>
                                                {pencilOutlineSvg}
                                            </a>
                                            {EntryDeleteBtn({
                                                ...this.props,
                                                entry,
                                                onSuccess: this.deleted,
                                                children: <a title={_t('g.delete')} className="delete-btn">{deleteForeverSvg}</a>
                                            })}
                                        </>
                                    )}
                                </div>
                            </>
                        })()}

                        {readMore && (
                            <div className="read-more">
                                {EntryLink({
                                    ...this.props,
                                    entry,
                                    children: <a>{_t("discussion.read-more")}</a>
                                })}
                            </div>
                        )}
                    </div>
                </div>

                {reply && Comment({
                    ...this.props,
                    defText: (ls.get(`reply_draft_${entry.author}_${entry.permlink}`) || ''),
                    submitText: _t('g.reply'),
                    cancellable: true,
                    onChange: this.replyTextChanged,
                    onSubmit: this.submitReply,
                    onCancel: this.toggleReply,
                    inProgress: inProgress,
                    autoFocus: true
                })}

                {edit && Comment({
                    ...this.props,
                    defText: entry.body,
                    submitText: _t('g.update'),
                    cancellable: true,
                    onSubmit: this.updateReply,
                    onCancel: this.toggleEdit,
                    inProgress: inProgress,
                    autoFocus: true
                })}

                {showSubList && <List {...this.props} parent={entry}/>}
            </div>
        );
    }
}


interface ListProps {
    history: History;
    location: Location;
    global: Global;
    dynamicProps: DynamicProps;
    users: User[];
    activeUser: ActiveUser | null;
    discussion: DiscussionType;
    parent: Entry;
    community: Community | null;
    ui: UI;
    addAccount: (data: Account) => void;
    setActiveUser: (username: string | null) => void;
    updateActiveUser: (data?: Account) => void;
    deleteUser: (username: string) => void;
    updateReply: (reply: Entry) => void;
    addReply: (reply: Entry) => void;
    deleteReply: (reply: Entry) => void;
    toggleUIProp: (what: ToggleType) => void;
}

export class List extends Component<ListProps> {
    render() {
        const {discussion, parent} = this.props;

        const {list} = discussion;
        let deletedPostID = ls.get('deletedComment')

        const filtered = list.filter(
            (x) => x.parent_author === parent.author && x.parent_permlink === parent.permlink && x.post_id !== deletedPostID
        );

        if (filtered.length === 0) {
            return null;
        }

        return (
            <div className="discussion-list">
                {filtered.map((d) => (
                    <Item key={`${d.author}-${d.permlink}`} {...this.props} entry={d}/>
                ))}
            </div>
        );
    }
}


interface Props {
    history: History;
    location: Location
    global: Global;
    dynamicProps: DynamicProps;
    users: User[];
    activeUser: ActiveUser | null;
    parent: Entry;
    community: Community | null;
    discussion: DiscussionType;
    ui: UI;
    addAccount: (data: Account) => void;
    setActiveUser: (username: string | null) => void;
    updateActiveUser: (data?: Account) => void;
    deleteUser: (username: string) => void;
    fetchDiscussion: (parent_author: string, parent_permlink: string) => void;
    sortDiscussion: (order: SortOrder) => void;
    resetDiscussion: () => void;
    updateReply: (reply: Entry) => void;
    addReply: (reply: Entry) => void;
    deleteReply: (reply: Entry) => void;
    toggleUIProp: (what: ToggleType) => void;
}

interface State {
    visible: boolean
}

export class Discussion extends Component<Props, State> {
    state: State = {
        visible: !!this.props.activeUser
    }

    componentDidMount() {
        const {activeUser} = this.props;
        if (activeUser) {
            this.fetch();
        }
    }

    componentDidUpdate(prevProps: Readonly<Props>): void {
        const {parent} = this.props;
        if (parent.url !== prevProps.parent.url) { // url changed
            this.fetch();
        }

        const {discussion} = this.props;
        if (prevProps.discussion.list.length === 0 && discussion.list.length > 0) {
            const {location} = this.props;
            if (location.hash) {
                const permlink = location.hash.replace("#", "");
                const anchorId = `anchor-${permlink}`;
                const anchorEl = document.getElementById(anchorId);
                if (anchorEl) {
                    anchorEl.scrollIntoView();
                }
            }
        }
    }

    componentWillUnmount() {
        const {resetDiscussion} = this.props;
        resetDiscussion();
    }

    fetch = () => {
        const {parent, fetchDiscussion} = this.props;
        const {author, permlink} = parent;
        fetchDiscussion(author, permlink);
    };

    orderChanged = (e: React.ChangeEvent<typeof FormControl & HTMLInputElement>) => {
        const order = e.target.value as SortOrder;
        const {sortDiscussion} = this.props;
        sortDiscussion(SortOrder[order]);
    };

    show = () => {
        this.setState({visible: true});
        this.fetch();
    }

    render() {
        const {parent, discussion, activeUser} = this.props;
        const {visible} = this.state;
        const {loading, order} = discussion;
        const count = parent.children;

        if (loading) {
            return <div className="discussion"><LinearProgress/></div>;
        }

        const join = <div className="discussion-card">
            <div className="icon">{commentSvg}</div>
            <div className="label">{_t("discussion.join")}</div>
            {LoginRequired({
                ...this.props,
                children: <Button>{_t("discussion.btn-join")}</Button>
            })}
        </div>;

        if (!activeUser && count < 1) {
            return <div className="discussion">{join}</div>;
        }

        if (count < 1) {
            return <div className="discussion empty"/>
        }

        const strCount = count > 1 ? _t("discussion.n-replies", {n: count}) : _t("discussion.replies");

        if (!visible && count >= 1) {
            return <div className="discussion">
                <div className="discussion-card">
                    <div className="icon">{commentSvg}</div>
                    <div className="label">{strCount}</div>
                    <Button onClick={this.show}>{_t("g.show")}</Button>
                </div>
            </div>
        }

        return (
            <div className="discussion">
                {!activeUser && <>{join}</>}
                <div className="discussion-header">
                    <div className="count"> {commentSvg} {strCount}</div>
                    <span className="flex-spacer"/>
                    <div className="order">
                        <span className="order-label">{_t("discussion.order")}</span>
                        <Form.Control as="select" size="sm" value={order} onChange={this.orderChanged} disabled={loading}>
                            <option value="trending">{_t("discussion.order-trending")}</option>
                            <option value="author_reputation">{_t("discussion.order-reputation")}</option>
                            <option value="votes">{_t("discussion.order-votes")}</option>
                            <option value="created">{_t("discussion.order-created")}</option>
                        </Form.Control>
                    </div>
                </div>
                <List {...this.props} parent={parent}/>
            </div>
        );
    }
}

export default (p: Props) => {
    const props: Props = {
        history: p.history,
        location: p.location,
        global: p.global,
        dynamicProps: p.dynamicProps,
        users: p.users,
        activeUser: p.activeUser,
        parent: p.parent,
        community: p.community,
        discussion: p.discussion,
        ui: p.ui,
        addAccount: p.addAccount,
        setActiveUser: p.setActiveUser,
        updateActiveUser: p.updateActiveUser,
        deleteUser: p.deleteUser,
        fetchDiscussion: p.fetchDiscussion,
        sortDiscussion: p.sortDiscussion,
        resetDiscussion: p.resetDiscussion,
        updateReply: p.updateReply,
        addReply: p.addReply,
        deleteReply: p.deleteReply,
        toggleUIProp: p.toggleUIProp
    }

    return <Discussion {...props} />;
}
