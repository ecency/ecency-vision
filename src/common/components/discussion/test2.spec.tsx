import React from 'react';

import Discussion from './index';

import { Discussion as DiscussionType, SortOrder } from '../../store/discussion/types';
import { allOver, UiInstance } from '../../helper/test-helper';

import { create, act } from 'react-test-renderer';

import { createBrowserHistory, createLocation } from 'history';

import {
  globalInstance,
  discussionInstace1,
  dynamicPropsIntance1,
  activeUserMaker
} from '../../helper/test-helper';

const [parent] = discussionInstace1;

const discussion: DiscussionType = {
  list: [],
  loading: false,
  error: false,
  order: SortOrder.trending
};

const defProps = {
  history: createBrowserHistory(),
  location: createLocation({}),
  global: globalInstance,
  dynamicProps: dynamicPropsIntance1,
  users: [],
  activeUser: null,
  parent: { ...parent, children: 0 },
  community: null,
  discussion,
  ui: UiInstance,
  addAccount: () => {},
  setActiveUser: () => {},
  updateActiveUser: () => {},
  deleteUser: () => {},
  fetchDiscussion: () => {},
  sortDiscussion: () => {},
  resetDiscussion: () => {},
  updateReply: () => {},
  addReply: () => {},
  deleteReply: () => {},
  toggleUIProp: () => {}
};

it('(1) Empty list with no active user', async () => {
  // render the component
  let component = await create(<Discussion {...defProps} />);

  // make assertions on component
  expect(component.toJSON()).toMatchSnapshot();
});

it('(2) Empty list with active user', async () => {
  const props = {
    ...defProps,
    activeUser: activeUserMaker('foo')
  };
  // render the component
  let component = await create(<Discussion {...props} />);

  // make assertions on component
  expect(component.toJSON()).toMatchSnapshot();
});
