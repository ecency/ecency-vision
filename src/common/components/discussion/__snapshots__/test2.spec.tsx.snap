// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(1) Empty list with no active user 1`] = `
<div
  className="discussion"
>
  <div
    className="discussion-card"
  >
    <div
      className="icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <path
          d="M18,8H6V6H18V8M18,11H6V9H18V11M18,14H6V12H18V14M22,4A2,2 0 0,0 20,2H4A2,2 0 0,0 2,4V16A2,2 0 0,0 4,18H18L22,22V4Z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      className="label"
    >
      Join the conversation now
    </div>
    <button
      className="cursor-pointer rounded-full duration-300 no-wrap bg-blue-dark-sky hover:bg-blue-dark-sky-hover focus:bg-blue-dark-sky-active text-white disabled:opacity-50 disabled:hover:bg-blue-dark-sky disabled:focus:bg-blue-dark-sky h-[2.125rem] px-3 "
      onClick={[Function]}
      style={
        Object {
          "outline": "none",
        }
      }
      type="button"
    >
      <div>
        Join
      </div>
    </button>
  </div>
</div>
`;

exports[`(2) Empty list with active user 1`] = `
<div
  className="discussion empty"
/>
`;
